clippy_lints/src/methods/parsed_string_literals.rs

```diff
@@ -0,0 +1,173 @@
+    static IPV4_ENTITY: &str = "an IPv4 address";
+    static IPV6_ENTITY: &str = "an IPv6 address";
+    let ipaddr_consts_available = || msrv.meets(cx, msrvs::IPADDR_CONSTANTS);
+    if is_type_diagnostic_item(cx, ty, sym::Ipv4Addr)
+        && let Some(sugg) = ipv4_subst(lit, ipaddr_consts_available())
+    {
+        maybe_emit_lint(
+            cx,
+            expr,
+            sugg.is_borrowed(),
+            IPV4_ENTITY,
+            sugg,
+            Applicability::MaybeIncorrect,
+        );
+    } else if is_type_diagnostic_item(cx, ty, sym::Ipv6Addr)
+        && let Some(sugg) = ipv6_subst(lit, ipaddr_consts_available())
+    {
+        maybe_emit_lint(
+            cx,
+            expr,
+            sugg.is_borrowed(),
+            IPV6_ENTITY,
+            sugg,
+            Applicability::MaybeIncorrect,
+        );
+    } else if is_type_diagnostic_item(cx, ty, sym::IpAddr) {
+        let with_consts = ipaddr_consts_available();
+        if let Some(sugg) = ipv4_subst(lit, with_consts) {
+            maybe_emit_lint(
+                cx,
+                expr,
+                sugg.is_borrowed(),
+                IPV4_ENTITY,
+                format!("IpAddr::V4({sugg})").into(),
+                Applicability::MaybeIncorrect,
+            );
+        } else if let Some(sugg) = ipv6_subst(lit, with_consts) {
+            maybe_emit_lint(
+                cx,
+                expr,
+                sugg.is_borrowed(),
+                IPV6_ENTITY,
+                format!("IpAddr::V6({sugg})").into(),
+                Applicability::MaybeIncorrect,
+            );
+        }
+    }
+}
+
+/// Suggests a replacement if `addr` is a correct IPv4 address
+fn ipv4_subst(addr: Symbol, with_consts: bool) -> Option<Cow<'static, str>> {
+    addr.as_str().parse().ok().map(|ipv4: Ipv4Addr| {
+        if with_consts && ipv4.as_octets() == &[127, 0, 0, 1] {
+            "Ipv4Addr::LOCALHOST".into()
+        } else if with_consts && ipv4.is_broadcast() {
+            "Ipv4Addr::BROADCAST".into()
+        } else if with_consts && ipv4.is_unspecified() {
+            "Ipv4Addr::UNSPECIFIED".into()
+        } else {
+            let ipv4 = ipv4.as_octets();
+            format!("Ipv4Addr::new({}, {}, {}, {})", ipv4[0], ipv4[1], ipv4[2], ipv4[3]).into()
+        }
+    })
+}
+
+/// Suggests a replacement if `addr` is a correct IPv6 address
+fn ipv6_subst(addr: Symbol, with_consts: bool) -> Option<Cow<'static, str>> {
+    addr.as_str().parse().ok().map(|ipv6: Ipv6Addr| {
+        if with_consts && ipv6.is_loopback() {
+            "Ipv6Addr::LOCALHOST".into()
+        } else if with_consts && ipv6.is_unspecified() {
+            "Ipv6Addr::UNSPECIFIED".into()
+        } else {
+            format!(
+                "Ipv6Addr::new([{}])",
+                ipv6.segments()
+                    .map(|n| if n < 2 { n.to_string() } else { format!("{n:#x}") })
+                    .join(", ")
+            )
+            .into()
```

Comment:
This doesn't have a test. `new` takes eight `u16`s, not an array.

Also same comment as before using `[..].into()` might be a better suggestion.